Add new tutorial: Web Scrape Amazon with Beautiful Soup (#10)
codedex-io/projects: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx

---
title: Web Scrape Amazon with Beautiful Soup
author: Ashutosh Krishna
datePublished: 2022-10-15
description: Learn how to web scrape Amazon using Beautiful Soup in this tutorial 
header: URL
tags:
  - beginner
  - python

>   - webscraping

Reviewer: Great job on the metadata! This is awesome. And let's take out the `webscraping` tag.

---

# Web Scrape Amazon with Beautiful Soup

<AuthorAvatar author_name="Ashutosh Krishna" author_avatar="/images/projects/authors/ashutosh_krishna.jpg" />

![Header image](URL)

> **Prerequisites:** Python fundamentals, Basics of HTML  

Reviewer 2: Can we keep this "Python fundamentals, HTML basics"? Just to maintain uniformity

> **Versions:** Python 3.10, requests 2.28.1, beautifulsoup4 4.11.1, lxml 4.9.1

Reviewer: ```suggestion
**Versions:** Python 3.10, requests 2.28.1, beautifulsoup4 4.11.1, lxml 4.9.1  
```

I added two spaces after so that Read Time go on a new line.

**Read Time:** 40 minutes  

## [#](#-introduction) Introduction

> Have you ever missed an Amazon product that was available at a discounted price? Have you ever thought of building an application that notifies you when the price of an Amazon product decreases below a certain price? If yes, let's build it together.

Reviewer: Like the hook!

> In this tutorial, we will learn to build a web scraper that scrapes Amazon product details using [Beautiful Soup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/).

Reviewer: I think we should explain Beautiful Soup (1-2 sentences) either here or in the **What is Web Scraping?** section?

The final result will look something like this:

> ![Demo](demo.gif)

Reviewer 2: Need to update this URL with the absolute URL from GitHub.

## [#](#-what-is-web-scraping) What is Web Scraping?

The process of extracting data from a website is known as web scraping. This data is collected and then exported in a way that the user will find more valuable, say an API or a CSV file. Although web scraping can be done manually, automated methods are typically preferable since they can be less expensive and perform more quickly.

While web scraping isn't illegal, it can become illegal if non-publicly available data is extracted. Web scraping can sometimes be difficult as the scraper may need to go through captchas or the website keeps constantly changing.

## [#](#-setting-up) Setting up

> First, open a Python code editor of your choice (VS Code recommended), and create a new Python file `scraper.py`. This is where we will be writing our code to scrape Amazon.

Reviewer: ```suggestion
First, open a Python code editor of your choice (VS Code recommended), and create a new Python file **scraper.py**. This is where we will be writing our code to scrape Amazon.
```

We haven't put this in the Contribution Guidelines yet, but file names on Codédex should be bold!

To start with, you will need to install three libraries: 
* The [`requests`](https://pypi.org/project/requests/) lets you send HTTP requests.
* The [`beautifulsoup4`](https://www.crummy.com/software/BeautifulSoup/bs4/doc/) lets you pull data out of HTML and XML files.
* The [`lxml`](https://www.crummy.com/software/BeautifulSoup/bs4/doc/#installing-a-parser) provides powerful API for parsing HTML and XML.

To install the above-mentioned libraries, run the below command in your VS Code terminal:
```bash
pip install requests beautifulsoup4 lxml
```

## [#](#-scraping-amazon) Scraping Amazon

Before we get started, let us import the libraries we will be using in our Python file:

```python
import requests
from bs4 import BeautifulSoup
```

> When we send an HTTP request, we send some **headers** along with the request. If you go to the Network tab of your browser and reload any web page, you can find the request headers similar to below:

Reviewer 2: The second sentence can sound confusing if someone who hasn't worked with DevTools is reading this article.

We can rephrase this as: "If you open devtools in your browsers(usually accessible through Command + Option + C or Control + Shift + C or F12 key) and switch to the network tab and reload any page, you can find the request headers that are similar to the ones shown below:"

Author: Yes @Goku-kun 
Thanks for the suggestion.
I was going to add it after @sonnynomnom asked to.

> ![Request Headers](request-headers.png)

Reviewer 2: Need to change this URL with the absolute URL.

Similar to browsers, we will also send some headers along with our HTTP request. Let's define them in a dictionary called `headers`:

```python
headers = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/106.0.0.0 Safari/537.36",
    "Accept-Language": "en-US,en;q=0.5"
}
```
To know your headers, you can also check out [this website](http://myhttpheader.com/).

Next, we will define a function `get_product_details()`. 

```python
def get_product_details(product_url: str) -> dict:
    pass
```
The function accepts a string parameter `product_url` and returns a dictionary. Let's call this dictionary `product_details`.

```python
def get_product_details(product_url: str) -> dict:

    # Create an empty product details dictionary
    product_details = {}

> ```

Reviewer 2: Need to link the Beautiful Soup documentation with details about the Soup object.

Now, let us make an HTTP request to get the HTML content of the product page. Using the `BeautifulSoup` class, we will create a soup of the page content.
```python
def get_product_details(product_url: str) -> dict:

    # Create an empty product details dictionary
    product_details = {}
    
    # Get the product page content and create a soup
    page = requests.get(product_url, headers=headers)
    soup = BeautifulSoup(page.content, features="lxml")
``` 

> Now comes the most important part of the tutorial where we will extract data from the page. For the sake of this tutorial, we will extract two data - the **title** of the product and the **price** of the product.

Reviewer: ```suggestion
Now comes the most important part of the tutorial where we will extract data from the page. For the sake of this tutorial, we will extract two data:

 - The **title** of the product.
 - The **price** of the product.
```

> ### Extracting the title of the product

Reviewer: Can we do title case for every header?

```suggestion
### Extracting the Title of the Product
```

> Open a product page, say [this one](https://www.amazon.com/Acer-Chromebook-Graphics-802-11ac-CB512-C1KJ/dp/B09Q9YKW15/ref=lp_16225007011_1_2) on your browser, and use the **Inspect Element** option to view the HTML content of the page. You will see the below HTML code for the title of the product.

Reviewer: Can we just show an image here of the Amazon Product page?

Reviewer: And then in the text, tell them how to get to the Inspect Elements. Like:

- On Windows: blah blah.
- On Mac: blah blah.

> Open a product page, say [this one](https://www.amazon.com/Acer-Chromebook-Graphics-802-11ac-CB512-C1KJ/dp/B09Q9YKW15/ref=lp_16225007011_1_2) on your browser, and use the **Inspect Element** option to view the HTML content of the page. You will see the below HTML code for the title of the product.

Reviewer 2: "Inspect Element" -> "Right click on the title of the page and select the 'Inspect Element' option from the dropdown menu"

> ![Extracting Title](extracting-title.png)

Reviewer 2: Replace the image link with the absolute URL.

The title is wrapped with a `span` tag with an attribute id **productTitle**. Let's use this id to extract the title of the product.

```python
title = soup.find("span", attrs={"id": "productTitle"}).get_text().strip()
```

> We use the `find()` method to find a `span` element. We pass the id **productTitle** in a dictionary called `attrs` that accepts the attributes. The `get_text()` method returns the text in a string format. The `strip()` method is used to remove any extra whitespaces.

Reviewer: ```suggestion
We use the `.find()` method to find a `span` element. We pass the `productTitle` id in a dictionary called `attrs` that accepts the attributes. The `.get_text()` method returns the text in a string format. The `.strip()` method is used to remove any extra whitespaces.
```

We haven't added in the Contribution Guide, the method names have the dot in front on Codédex.

> We use the `find()` method to find a `span` element. We pass the id **productTitle** in a dictionary called `attrs` that accepts the attributes. The `get_text()` method returns the text in a string format. The `strip()` method is used to remove any extra whitespaces.

Reviewer 2: The `strip()` method is used to remove any extra leading and trailing whitespaces.

### Extracting the price of the product

> Similar to the title of the product, if you inspect the page, you will find the below HTML code for the price of the product.

Reviewer 2: Similar to the title of the product, if you inspect the prices by right clicking on them and opening devtools, you will find the below HTML code for the price of the product.

![Extracting Price](extracting-price.png)

Thus, we can extract the price in a similar fashion.

```python
price = soup.find("span", attrs={"class": "apexPriceToPay"}).get_text().strip()
```

But, you will see a problem when you print the price of the product. The extracted price will be something like **$166.00$166.00** because the parent `span` element contains two `span` elements with the price text in them. But we can clean this extracted price to get the price of the product in the following way:

```python
extracted_price = soup.find("span", attrs={"class": "apexPriceToPay"}).get_text().strip()
price = "$" + extracted_price.split("$")[1]
```

> We split the `extracted_price` string using the **$** symbol and select the element at index 1. We also add the dollar symbol before the price.

Reviewer: ```suggestion
We split the `extracted_price` string using the `$` symbol and select the element at index 1. We also add the dollar symbol before the price.
```

Also should we briefly (in a few words) explain the $ sign?

        extracted_price = soup.find("span", attrs={"class": "apexPriceToPay"}).get_text().strip()
        price = extracted_price.split("$")[1]
        
        
        # Adding it to the product details dictionary
        product_details["title"] = title
        product_details["price"] = price
        product_details["product_url"] = product_url

        # Return the product details dictionary
        return product_details
    except Exception as e:
        print("Could not fetch product details")
        print(f"Failed with exception: {e}")
```

Notice that we have used **try-except** block to catch any error.

## [#](#-running-the-scraper) Running the scraper

To run the scraper, we will ask the user to enter any product URL. Then we will call the `get_product_details()` function with that product URL and print it.

```python
product_url = input("Enter product url: ")
product_details = get_product_details(product_url)

print(product_details)

```

## [#](#-conclusion) Conclusion

In this tutorial, we scraped Amazon to extract the title and price of a product. You can follow the same pattern to extract more details like the rating of the product or its availability.

> As mentioned before, the layout of a website keeps changing and hence, the above code may not be helpful in the future. But this tutorial should be helpful to teach to you how to scrape any website.

Reviewer 2: "the above code may not be helpful" -> "the above code may no longer be applicable as it is, without some changes. Join Codédex club to get support from code mentors if you run any trouble while using this article."

@sonnynomnom What do you think about this? Just the second updated sentence.

> As mentioned before, the layout of a website keeps changing and hence, the above code may not be helpful in the future. But this tutorial should be helpful to teach to you how to scrape any website.

Reviewer: Yeah we can do that!

```suggestion
As mentioned before, the layout of Amazon will keep changing and hence, the above code may not be helpful in the future. Join Codédex Club to get support from code mentors if you run any trouble while using this article!
```

datePublished: 2022-10-15
description: Learn how to web scrape Amazon using Beautiful Soup in this tutorial 
header: URL
tags:
  - beginner
  - python
---

# Web Scrape Amazon with Beautiful Soup

<AuthorAvatar author_name="Ashutosh Krishna" author_avatar="/images/projects/authors/ashutosh_krishna.jpg" />

![Header image](URL)

**Prerequisites:** Python fundamentals, HTML basics  
**Versions:** Python 3.10, requests 2.28.1, beautifulsoup4 4.11.1, lxml 4.9.1  
**Read Time:** 40 minutes  

## [#](#-introduction) Introduction

Have you ever missed an Amazon product that was available at a discounted price? Have you ever thought of building an application that notifies you when the price of an Amazon product decreases below a certain price? If yes, let's build it together.

In this tutorial, we will learn to build a web scraper that scrapes Amazon product details using [Beautiful Soup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/).

The final result will look something like this:

![Demo](https://raw.githubusercontent.com/codedex-io/projects/main/projects/web-scrape-amazon-with-beautiful-soup/demo.gif)

## [#](#-what-is-web-scraping) What is Web Scraping?

> The process of extracting data from a website is known as web scraping. This data is collected and then exported in a way that the user will find more valuable, say an API or a CSV file. Although web scraping can be done manually, automated methods are typically preferable since they can be less expensive and perform more quickly.

Reviewer 3: A Hyperlink to this wiki article might be helpful for users wanting to get a deeper understanding of web scraping.

 https://en.wikipedia.org/wiki/Web_scraping

![Request Headers](https://raw.githubusercontent.com/codedex-io/projects/main/projects/web-scrape-amazon-with-beautiful-soup/request-headers.png)

Similar to browsers, we will also send some headers along with our HTTP request. Let's define them in a dictionary called `headers`:

```python
headers = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/106.0.0.0 Safari/537.36",
    "Accept-Language": "en-US,en;q=0.5"
}
```
To know your headers, you can also check out [this website](http://myhttpheader.com/).

Next, we will define a function `get_product_details()`. 

```python
def get_product_details(product_url: str) -> dict:
    pass
```
The function accepts a string parameter `product_url` and returns a dictionary. Let's call this dictionary `product_details`.

```python
def get_product_details(product_url: str) -> dict:

    # Create an empty product details dictionary
    product_details = {}
```

> Now, let us make an HTTP request to get the HTML content of the product page. Using the `BeautifulSoup` class, we will create a [soup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/#making-the-soup) of the page content.

Reviewer 3: For me, "let us" breaks the flow of the sentence. "Now, let's make" would sound better I think.
This is subjective of course.

```python
def get_product_details(product_url: str) -> dict:

    # Create an empty product details dictionary
    product_details = {}
    
    # Get the product page content and create a soup
    page = requests.get(product_url, headers=headers)
    soup = BeautifulSoup(page.content, features="lxml")
``` 

Now comes the most important part of the tutorial where we will extract data from the page. For the sake of this tutorial, we will extract two data: 

- The **title** of the product.
- The **price** of the product.

> But before extracting, let us learn how to view the HTML code that makes up any web page. Open a product page, say [this one](https://www.amazon.com/Acer-Chromebook-Graphics-802-11ac-CB512-C1KJ/dp/B09Q9YKW15) on your browser. 

Reviewer 3: "let us" --> "let's"

![Amazon Product Page](https://raw.githubusercontent.com/codedex-io/projects/main/projects/web-scrape-amazon-with-beautiful-soup/amazon-product-page.png)

To view the HTML code, we have an **Inspect** option in every browser. You can use the keyboard shortcuts – Ctrl + Shift + I for Windows or Linux and Cmd + Option + I for macOS users. Alternatively, right-click on the web page and choose **Inspect** to access the Developer tools panel. Make sure you are on the **Elements** tab in the Developer tools panel.

![Inspect Element](https://raw.githubusercontent.com/codedex-io/projects/main/projects/web-scrape-amazon-with-beautiful-soup/inspect-element.gif)

### Extracting the Title of the Product
 
Use the **Inspect** option to view the HTML code used for the title of the product.

![Extracting Title](https://raw.githubusercontent.com/codedex-io/projects/main/projects/web-scrape-amazon-with-beautiful-soup/extracting-title.png)

The title is wrapped with a `span` tag with an attribute id **productTitle**. Let's use this id to extract the title of the product.

```python
title = soup.find("span", attrs={"id": "productTitle"}).get_text().strip()
```

> We use the `.find()` method to find a `span` element. We pass the `productTitle` id in a dictionary called `attrs` that accepts the attributes. The `.get_text()` method returns the text in a string format. The `.strip()` method is used to remove any extra leading and trailing whitespaces.

Reviewer 3: Replace "we" with "Then" in the second sentence.

Author: Done @asiqurrahman

        extracted_price = soup.find("span", attrs={"class": "apexPriceToPay"}).get_text().strip()
        price = extracted_price.split("$")[1]
        
        
        # Adding it to the product details dictionary
        product_details["title"] = title
        product_details["price"] = price
        product_details["product_url"] = product_url

        # Return the product details dictionary
        return product_details
    except Exception as e:
        print("Could not fetch product details")
        print(f"Failed with exception: {e}")
```

Notice that we have used **try-except** block to catch any error.

## [#](#-running-the-scraper) Running the Scraper

To run the scraper, we will ask the user to enter any product URL. Then we will call the `get_product_details()` function with that product URL and print it.

```python
product_url = input("Enter product url: ")
product_details = get_product_details(product_url)

print(product_details)

```

## [#](#-conclusion) Conclusion

In this tutorial, we scraped Amazon to extract the title and price of a product. You can follow the same pattern to extract more details like the rating of the product or its availability.

> As mentioned before, the layout of Amazon will keep changing and hence, the above code may not be helpful in the future. Join Codédex Club to get support from code mentors if you run any trouble while using this article!

Reviewer 3: "if you run any trouble" should be "if you run into any trouble".

That should be my final feedback.
Amazing article, loved the pace of it and how easy it is to understand!

Author: Great, thank you @asiqurrahman 
I've made the required changes.
